### DISCUSSION TITLE
2024-09-06 - GitHub - Reviewing PRs - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by PothieuG** [SSWConsulting/SSW.Rules.Content#9109] [Chewing the Fat]
Hi PR reviewers,
https://www.ssw.com.au/rules/pr-suggest-changes/

Did you already know about the GitHub "Add a suggestion" button?
What are your best tips for reviewing PRs?

**Comment by tiagov8** [comment]
I didn't know about this yet. It's looks the most efficient way to provide some extra feedback to a PR.

I always check for more things than what was changed when reviewing PRs for Rules. I expect to be using this feature quite a lot.

**Comment by brookjeynes-ssw** [comment]
The leave a suggestion feature is awesome for the person getting their pr reviewed as it makes the process 10x easier to make changes for the reviewer.

**Comment by danielmackay** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes, I've used this a bunch.  However, I find it only works well for textual changes or simple code changes.  I certainly wouldn't refactor a large method using this code.

> What are your best tips for reviewing PRs?

When reviewing pull request comments can often be divided into three categories.  I like to use prefixes to differentiate between these:
 
- **Change** - you must change this if you want me to approve.
- **Consider** – consider this suggestion, but don’t feel obliged to change (i.e. I would do this, but I won’t block you on it)
- **PBI** – Important, but create a PBI to fix this in future

**Comment by TobyChurches** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes - I mainly use DevOps at SSW but it has a similar feature. Super handy especially when the change isn't particularly large it can speed up the review process a lot.

> What are your best tips for reviewing PRs?

If possible always use a suggestion + always add screenshots of UI changes in the Pr description (its too hard to visualise CSS changes)

**Comment by leahy268** [comment]
>Did you already know about the GitHub "Add a suggestion" button?

Yes, Wicksy showed it to me over a year ago, been using it ever since as it's awesome for the reviewee.  

>What are your best tips for reviewing PRs?

Use suggestions wherever you can and explain yourself well.  If you can't explain them well or there are a lot of comments, give the other person a quick call and take them through it.

**Comment by bradystroud** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes and its great for little suggestions

> What are your best tips for reviewing PRs?

AFTER your review, paste the diff into ChatGPT and see if it catches anything you didn't

**Comment by jakebayliss** [comment]
What about DevOps?
Not sure how helpful this will be if it is actually code and not just content. Small changes easy but other than content and tiny changes e.g. variable names or something, might not be that useful. 
Would prefer reviewer to ask questions about my code so that I can think about what I am doing, not just tell me exactly what should change

**Reply by wicksipedia** [reply]
AzDO has the same feature

**Comment by ncn-ssw** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

✅ Yes

> What are your best tips for reviewing PRs?

- Execute the changed code - it is stunning the number of times that a developer just assumes that their change would work, or that their build system has cached pieces of a previous build
- Clear your own caches before testing - ensure you are actually testing the change
- Look at the conditionals affected by the change - can you manipulate the input data to trigger all paths?
- Don't expect a PR to solve all the world's problems as it is just a finite piece of work output - but if there are follow up issues, ensure that they are documented

**Comment by uly1** [comment]
I strongly dislike the requirement for a PR for content changes. It adds layers of beaurocracy to what should be a simple and streamlined process. 
e.g. adding punctuation to a sentence or a spelling mistake

As a minimum, there should be a like of trusted people who are able to make website changes without a PR. e.g. Managers, Designers, Marketers, TV people, etc.

We'd still have version history, and be able to roll back or see who made what change. 

PLEASE remove the PRs from content changes!!!

**Reply by wicksipedia** [reply]
![computer-gif](https://github.com/user-attachments/assets/5c47afc0-a873-43fb-88b1-6c01edb089d5)

**Reply by PennyWalker** [reply]
I have to agree with this one, it's annoying to have to wait to change a spelling error or change a single word because no one is free to review your PR.

**Comment by tkapa** [comment]
I already knew about this feature, and it also exists in AzDo for you old heads.

I +1 Dan's comment, there can be changes that aren't absolutely necessary to push the PR, and it should be clear to the reader which changes are necessary vs considerations.

**Comment by ricksu978** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes

> What are your best tips for reviewing PRs?

- Make sure PR has linked to the issue it's resolving
- Check the Acceptance Criteria

**Comment by GordonBeeming** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes

> What are your best tips for reviewing PRs?

Don't rush them... a rushed PR review is worse than no PR review

**Reply by ozdemir-mehmet** [reply]
Suggestion button looks good, I will probably use it going forward for my detailed suggestions

**Comment by edkranz** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes

> What are your best tips for reviewing PRs?

Don't hold back

**Comment by StefStar** [comment]
I did not know about the suggestion button in Github. 

Being new to GitHub and only have reviewed one PR, the more obvious the proposed changes are the easier it is!

**Comment by isaaclombardssw** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yep - I'm a fan.

> What are your best tips for reviewing PRs?

Probe any of it you don't understand, might learn something too.

**Comment by Calinator444** [comment]
Hi PR reviewers,
https://www.ssw.com.au/rules/pr-suggest-changes/

> Did you already know about the GitHub "Add a suggestion" button?

Yes - I try to unblock people quickly when I'm asked to review a PR and adding code suggestions helps developers address feedback quicker.

> What are your best tips for reviewing PRs?

If the change you're requesting is systematic it may be best to call the PR owner, explain the changes you want, and then add an "as per our conversation" comment in the PR before requesting changes.

**Comment by Tanya-Leahy** [comment]
>Did you already know about the GitHub "Add a suggestion" button?

No, but I know how to use it now!
>What are your best tips for reviewing PRs?

No tips unfortunately as I haven't had to review PRs.

**Comment by piers-sinclair** [comment]
I know the feature and I use it all the time, super handy! 

One other tip is to only select the lines you want to make a comment on, many people don't know this is possible in GitHub but it will save a lot of time and reduce confusion for reviewers.

Here's a nice GIF which shows it:
[https://x.com/natfriedman/status/1179097330097643521](javascript:void(0);)

**Comment by jaydenalchin** [comment]
I know and love the GitHub suggestions feature, it makes feedback quick & easy.

> What are your best tips for reviewing PRs?

Include before and after screenshots in PR description when there are UI changes, which helps all the visual creatures (like me) get quick context and understanding around the content we're reviewing.

**Comment by amankumarrr** [comment]
Yes, I recently discovered this feature, and it's really useful, especially for content changes as it reduces the margin of error. However, for code suggestions, it can be a bit tricky since you need to test the changes first before committing the suggestion.

**Comment by bahjatalaadelssw** [comment]
I didn't know about it, but I like how it generates a clear visual a/b difference to speed up the editing process.

**Comment by wicksipedia** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

https://youtu.be/xPFe3MlOJQ4?si=cb6g4d6EqwSbNDJ7&t=1631

> What are your best tips for reviewing PRs?

it doesn't have to be perfect - it just has to be better
you need to merge the PR in order to ship to learn
move fast, fix things

**Comment by jeoffreyfischer** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

No I didn't - I will definitely use it now! It's a game changer 💪

> What are your best tips for reviewing PRs?

One word: SMALLER - many PRs are too big which makes the reviewing process painful and less efficient. There's a huge difference in the review outcome between a PR with 3 file changes and a PR with 100 file changes.

**Comment by AttackOnMorty** [comment]
This feature is good for content change. But not suitable for code changes, you might need to update the unit tests and test the changes locally.

**Reply by wicksipedia** [reply]
Disagree, I think it is really useful with code changes. It's often easier to explain the change by actually doing it, than explaining it.

You can always update your branch with the changes on remote and update your unit tests afterwards

**Reply by danielmackay** [reply]
Depends on the size of the code change.  Personally, I wouldn't feel comfortable suggestion large code changes via this feature without intellisense, being able to compile, etc.

**Comment by awaltos** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yep

> What are your best tips for reviewing PRs?

Seeing changes on github.com for small PRs are so nice. I think they have a very decent UX.

To review larger PRs I like to do this in Rider or VSC using the GitHub plugins/extensions. This allows me to see the change in context and I can click through or hover definitions of things I don't understand. The VSC GH Pull Requests extension also allows us to add suggestions as reviewers and decide to include theses suggestions as requesters - this is a great experience!

**Comment by RobJT-wq** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

No

> What are your best tips for reviewing PRs?

Making the PR clear and concise for the reviewer

**Comment by chrisschultzssw** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yep

> What are your best tips for reviewing PRs?

ship it ¯\\_(ツ)_/¯

**Reply by RobJT-wq** [reply]
(╯°□°）╯︵ ┻━┻

**Comment by bettybondoc** [comment]
Brady actually introduced me to this feature months ago. I think it's a game changer!

**Comment by MattParkerDev** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes.

> What are your best tips for reviewing PRs?

If you have a lot of feedback, its usually better to call the person and go through their PR with them.

**Comment by JackDevAU** [comment]
Anything that makes the person(s) reviewing our code easier should be done. "Code reviewers are our first users—make their job easy, and the whole team wins."

Suggestions, good descriptions with images and list of fixes, features, etc + link to an issue is perfect 🙏
Hot tip - Review your own PR! 
Hot tip #2 - Ask questions when reviewing PRs to make your feedback sound less "mean"

**Reply by bradystroud** [reply]
> Hot tip - Review your own PR!

This is great

**Comment by jimmidier** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes I did.

> What are your best tips for reviewing PRs?

As the reviewer, prepend severity would help the author better understand the feedback. As the author, call the reviewer and go through the PR together if it's a large one.

**Comment by jerryluossw** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

No, but I belive I will use it next time.

> What are your best tips for reviewing PRs?

The suggestions must be detailed and clear enough. If the logic and reasons behind the suggestions are very complex, it's best to include a Teams call.

**Comment by babakamyljanovssw** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Not yet

> What are your best tips for reviewing PRs?

Do over shoulder review when PR is big

**Comment by KristenHu** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes.
> What are your best tips for reviewing PRs?

It's better to have a call if there are some big changes

**Comment by Freego1783** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yeah, I use it for content changes and small code changes!

> What are your best tips for reviewing PRs?

If you're adding a lot of comments on a PR, don't hesitate to call the PR owner to go through it with him

**Comment by tino-liu** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes.

> What are your best tips for reviewing PRs?

- Make sure it runs for both happy and unhappy paths
- If there are a big batch of suggestions, I'd rather call, then later add a `as per my conversation...` comment and summarize the changing points.

**Comment by duanxinhuan** [comment]
Each team has its own PR review process, they can be inefficient or efficient, that's up to them. All we need to do is it give them suggestions to get better like this CTF.
But enforce efficiency will be a disaster because human beings is kind of not good at being enforced.

**Comment by duanxinhuan** [comment]
Also I found vs code github extension do not have suggestion feature:
![image](https://github.com/user-attachments/assets/1445889d-8333-4b0f-bc28-dd5335d25b46)
we should fix this by asking github team to fix

**Comment by jimmy1992abc** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

No - First time, it looks good and should be the default for any changes in GitHub (without the button)

> What are your best tips for reviewing PRs?

For small fix, I think what GitHub provides is already good enough.
For big fix, review the PR with the contributors together in a call.

**Comment by ozair-ashfaque** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes

> What are your best tips for reviewing PRs?

Use of prefixes in the code review:
[Do you use prefixes to improve code review communication?](https://www.ssw.com.au/rules/use-prefixes-to-improve-code-review-communication)

**Comment by vladislav-kir** [comment]
Add a Suggestion is a great feature, that makes the process significantly quicker

**Comment by Ben0189** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

Yes i already know this

> What are your best tips for reviewing PRs?

- PR reviewer should not leave too much comment at a time (the better way is to address the more important things at the first, and then request more changes afterward once they resolve the first time comment)
- PR comment for code improvement should have some link to external resources so it help pr owner to learn on what they can improve
- Sometimes the pr comment hard to explain / understand, then this is a good sign for the pr reviewer to call the pr owner for explanation

**Comment by PennyWalker** [comment]
I didn't already know about this feature, but I so rarely use Github or make huge changes. I can see why it would be helpful for devs.

**Comment by Marxoz** [comment]
> Did you already know about the GitHub "Add a suggestion" button?

No but it is cool for text changes

> What are your best tips for reviewing PRs?

Always make sure you understand the context before reviewing!
But also make sure you explain any suggestions thoroughly

**Comment by calumjs** [comment]
Before asking someone else to review your PR, try copy pasting your PR changes into chatgpt and ask it for suggestions. Hopefully this will help pick off some low hanging fruit

**Comment by stephanfakossw** [comment]
That's a nice way to change PR text :-)

> Did you already know about the GitHub "Add a suggestion" button?  

Nope, but looks cool

> What are your best tips for reviewing PRs?  

Not having much opportunities to review PR's as an accountant
